File: doubly_linked_lists/ruby/doubly_linked_list.rb
require_relative 'node'

class DoublyLinkedList
  attr_reader :head, :tail, :length

  def initialize(value)
    first_node = Node.new(value)
    @head = first_node
    @tail = first_node
    @length = 1
  end

  def print_list
    curr = @head
    until curr.nil?
      puts curr.value
      curr = curr.next
    end
  end

  def append(value)
    new_node = Node.new(value)
    if !@head && !@tail
      @head = new_node
    else
      @tail.next = new_node
      new_node.prev = @tail
    end
    @tail = new_node
    @length += 1
    new_node
  end

  def pop
    return unless @head && @tail

    last_node = @tail
    if @head.equal?(@tail) # https://shopify.engineering/implementing-equality-in-ruby
      @tail = nil
      @head = nil
    else
      @tail = last_node.prev
      @tail.next = nil
      last_node.prev = nil
    end
    @length -= 1
    last_node
  end

  def prepend(value)
    new_node = Node.new(value)
    if !@head && !@tail
      @head = new_node
      @tail = new_node
    else
      new_node.next = @head
      @head.prev = new_node
      @head = new_node
    end
    @length += 1
    new_node
  end

  def pop_first
    return unless @head && @tail

    first_node = @head
    if @head.equal?(@tail) # https://shopify.engineering/implementing-equality-in-ruby
      @head = nil
      @tail = nil
    else
      @head = first_node.next
      @head.prev = nil
      first_node.next = nil
    end
    @length -= 1
    first_node
  end

  def get(index)
    return unless @head && @tail && index.between?(0, @length - 1)

    if index.between?(0, @length.fdiv(2).floor)
      i = 0
      curr = @head
      until i == index
        curr = curr.next
        i += 1
      end
    else # index.between?(@length.fdiv(2).ceil, @length - 1)
      i = @length - 1
      curr = @tail
      until i == index
        curr = curr.prev
        i -= 1
      end
    end
    curr
  end

  def set_value(index, value)
    node = get(index)
    return false unless node

    node.value = value
    true
  end

  def insert(index, value)
    return false unless index.between?(0, @length - 1)

    if index.zero?
      prepend(value)
    elsif index == @length
      append(value)
    else
      new_node = Node.new(value)
      before = get(index - 1)
      after = before.next
      new_node.next = after
      after.prev = new_node
      before.next = new_node
      new_node.prev = before
    end
    @length += 1
    true
  end

  def remove(index)
    return nil unless index.between?(0, @length - 1)

    if index.zero?
      node = pop_first
    elsif index == @length - 1
      node = pop
    else
      node = get(index)
      node.next.prev = node.prev
      node.prev.next = node.next
      node.next = nil
      node.prev = nil
      node
    end
    @length -= 1
    node
  end
end

File: doubly_linked_lists/python/doubly_linked_list.py
from node import Node

class DoublyLinkedList: # https://www.geeksforgeeks.org/doubly-linked-list/
    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0

    def print_list(self):
        curr = self.head
        while curr: # loop until curr is None
            print(curr.value, end=" -> ")
            curr = curr.next
        print("None")

    def prepend(self, value): # add at head
        new_node = Node(value)
        if not self.head: # Is empty
            self.head = self.tail = new_node
        else:
            new_node.next = self.head
            self.head.prev = new_node
            self.head = new_node
        self.size += 1

    def append(self, value): # add at tail
        new_node = Node(value)
        if not self.tail:
            self.head = self.tail = new_node
        else:
            new_node.prev = self.tail
            self.tail.next = new_node
            self.tail = new_node
        self.size += 1

    def insert(self, index, value): # add at index
        if index < 0 or index > self.size:
            return
        if index == 0:
            self.prepend(value)
            return
        if index == self.size:
            self.append(value)
            return

        new_node = Node(value)
        curr = self.head
        for _ in range(index - 1):
            curr = curr.next

        new_node.next = curr.next
        new_node.prev = curr
        curr.next.prev = new_node # adjust the next node's prev pointer
        curr.next = new_node # adjust the current node's next pointer
        self.size += 1

    def pop_first(self): # delete at head
        if not self.head: # empty list
            return
        if self.head == self.tail: # only one node
            self.head = self.tail = None
        else:
            self.head = self.head.next
            self.head.prev = None
        self.size -= 1

    def pop_last(self): # delete at tail
        if not self.tail:
            return
        if self.head == self.tail:
            self.head = self.tail = None
        else:
            self.tail = self.tail.prev
            self.tail.next = None
        self.size -= 1

    def remove(self, index):
        if index < 0 or index > self.size:
            return
        if index == 0:
            self.pop_first()
            return
        if index > self.size:
            self.pop_last()
            return

        curr = self.head
        for _ in range(index):
            curr = curr.next

        curr.prev.next = curr.next # adjust the prev node's next pointer
        if curr.next:
            curr.next.prev = curr.prev # adjust the next node's prev pointer
        self.size -= 1

    def get(self, index):
        if index < 0 or index >= self.size: # validate index
            return -1

        if index < self.size // 2: # 1st half of the list
            curr = self.head
            for _ in range(index):
                curr = curr.next
        else: # 2nd half of the list
            curr = self.tail
            for _ in range(self.size - 1, index, -1):
                curr = curr.prev

        return curr.value
